File: frontend/src/api/m3.ts
import { apiClient } from "./client";

export type WikiPageType = "index" | "source" | "entity" | "concept" | "overview" | "analysis";

export type WikiPageSummary = {
  id: string;
  kb_id: string;
  slug: string;
  title: string;
  page_type: WikiPageType;
  summary: string;
  category_path: string[];
  aliases: string[];
  source_refs: string[];
  updated_at: string;
};

export type WikiPage = WikiPageSummary & {
  content: string;
  current_revision_id: string;
  manual_edit_warning?: boolean;
  created_at: string;
};

export type WikiPageTreeNode = {
  name: string;
  path: string[];
  pages: WikiPageSummary[];
  children: WikiPageTreeNode[];
};

export type WikiGraphNode = {
  id: string;
  name: string;
  slug: string;
  entity_type: string;
  wiki_page_id?: string | null;
};

export type WikiGraphEdge = {
  id: string;
  source_entity_id: string;
  target_entity_id: string;
  relation_type: string;
  source_chunk_id?: string | null;
};

export type WikiGraph = {
  nodes: WikiGraphNode[];
  edges: WikiGraphEdge[];
};

export type WikiPageSourceChunk = {
  id: string;
  seq: number;
  header_path: string[];
  content: string;
  start_pos: number;
  end_pos: number;
};

export type WikiPageSource = {
  document_id: string;
  filename: string;
  status: string;
  precise: boolean;
  chunks: WikiPageSourceChunk[];
};

export async function ingestWiki(kbId: string, documentIds?: string[]): Promise<{ task_id: string }> {
  const response = await apiClient.post<{ task_id: string }>(`/wiki/${kbId}/ingest`, {
    document_ids: documentIds,
  });
  return response.data;
}

export async function rebuildWiki(kbId: string): Promise<{ task_id: string }> {
  const response = await apiClient.post<{ task_id: string }>(`/wiki/${kbId}/rebuild`, {
    confirm: "REBUILD",
  });
  return response.data;
}

export async function listWikiPages(params: {
  kbId: string;
  q?: string;
  page_type?: WikiPageType | "";
}): Promise<{ items: WikiPageSummary[]; tree: WikiPageTreeNode[] }> {
  const { kbId, ...query } = params;
  const response = await apiClient.get<{ items: WikiPageSummary[]; tree: WikiPageTreeNode[] }>(`/wiki/${kbId}/pages`, {
    params: query,
  });
  return response.data;
}

export async function getWikiPage(pageId: string): Promise<WikiPage> {
  const response = await apiClient.get<WikiPage>(`/wiki-pages/${pageId}`);
  return response.data;
}

export async function getWikiPageSources(pageId: string): Promise<WikiPageSource[]> {
  const response = await apiClient.get<{ items: WikiPageSource[] }>(`/wiki-pages/${pageId}/sources`);
  return response.data.items;
}

export async function getWikiGraph(params: {
  kbId: string;
  entity_type?: string;
  relation_type?: string;
}): Promise<WikiGraph> {
  const { kbId, ...query } = params;
  const response = await apiClient.get<WikiGraph>(`/wiki/${kbId}/graph`, { params: query });
  return response.data;
}

File: backend/app/schemas/m3.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

WikiPageType = Literal["index", "source", "entity", "concept", "overview", "analysis"]
EditorType = Literal["agent", "manual"]


class WikiIngestRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    document_ids: list[str] | None = None


class WikiRebuildRequest(BaseModel):
    confirm: Literal["REBUILD"]


class WikiPageSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    kb_id: str
    slug: str
    title: str
    page_type: WikiPageType
    summary: str
    category_path: list[str]
    aliases: list[str]
    source_refs: list[str]
    updated_at: datetime


class WikiPageOut(WikiPageSummary):
    content: str
    current_revision_id: str
    manual_edit_warning: bool = False
    created_at: datetime


class WikiPageSourceChunk(BaseModel):
    id: str
    seq: int
    header_path: list[str]
    content: str
    start_pos: int
    end_pos: int


class WikiPageSource(BaseModel):
    document_id: str
    filename: str
    status: str
    precise: bool
    chunks: list[WikiPageSourceChunk]


class WikiPageSourceResponse(BaseModel):
    items: list[WikiPageSource]


class WikiPageTreeNode(BaseModel):
    name: str
    path: list[str]
    pages: list[WikiPageSummary]
    children: list["WikiPageTreeNode"]


class WikiPageListResponse(BaseModel):
    items: list[WikiPageSummary]
    tree: list[WikiPageTreeNode]


class WikiPageUpdateRequest(BaseModel):
    content: str = Field(min_length=1)
    change_summary: str | None = Field(default=None, max_length=512)


class WikiPageRevisionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    page_id: str
    content: str
    editor_type: EditorType
    editor_id: str | None = None
    change_summary: str
    created_at: datetime


class WikiRevisionPage(BaseModel):
    items: list[WikiPageRevisionOut]
    total: int
    page: int
    page_size: int


class WikiGraphNode(BaseModel):
    id: str
    name: str
    slug: str
    entity_type: str
    wiki_page_id: str | None = None


class WikiGraphEdge(BaseModel):
    id: str
    source_entity_id: str
    target_entity_id: str
    relation_type: str
    source_chunk_id: str | None = None


class WikiGraph(BaseModel):
    nodes: list[WikiGraphNode]
    edges: list[WikiGraphEdge]
